### ISSUE TITLE
Bootcamp curriculum discussion

### ISSUE THREAD

**Issue by dominguezmaria** [LDSSA/curriculum-development#33]
We'll use this issue to discuss changes to the [bootcamp curriculum](https://github.com/LDSSA/curriculum-development/blob/master/curriculum/01-bootcamp-and-binary-classification.md).

I'm proposing to remove the SLUs about ML models (other than Linear and Logistic Regression) and include them as optional content in Specializations:
- SVMs would go to Text Classification
- kNN would go to RecSys
- Tree Based models would still be in the Bootcamp (but as optional content).

Optional content means that:
- we won't present it in the bootcamp, i.e, students should go over it by themselves
- the exercises are not mandatory
- students are encouraged to study this content and they can ask questions about it 

The reasons for this proposal are:
- the bootcamp is very heavy
- if we present the models together with a more real-world example of its usage, it's more likely that the students will understand it
- by making it optional, we leave it up to the students to decide if they have the capacity, at that moment, to absorb that information, or if they prefer to do it later

**Comment by dominguezmaria** [comment]
Questions for @CatarinaSilva:
- When do we need the final version of the bootcamp curriculum by, so we can add a due date to the issue
- Could you please add here the suggestions that you mentioned?

**Comment by PedroGFonseca** [comment]
I like the idea of simplifying, but we may want to keep the tree based models as compulsory because￼ they
- are intuitive 
- form the basis for the random forest "pocket knife"
- are useful to teach overfitting

**Comment by dominguezmaria** [comment]
> I like the idea of simplifying, but we may want to keep the tree based models as compulsory because￼ they
> 
> * are intuitive
> * form the basis for the random forest "pocket knife"
> * are useful to teach overfitting

That's fair!
It also allows to keep the SLU about hyperparameter tuning almost unchanged, since a lot of the examples were on tree based methods.

**Comment by Juliana-R** [comment]
My 2 cents: consider the time effort reported by more than 30 students was significantly more than the 10 hours/week recommended by us.

https://github.com/LDSSA/wiki/issues/64

**Comment by Juliana-R** [comment]
Consider @PedroGFonseca 's suggestions: https://github.com/LDSSA/wiki/issues/75.

**Comment by dominguezmaria** [comment]
> Consider @PedroGFonseca 's suggestions: [LDSSA/wiki#75](https://github.com/LDSSA/wiki/issues/75).

I think these are teaching concerns rather than curriculum concerns. As in, these are about how to organise the bootcamp and not what content should go into the bootcamp.
I'm ok with the bootcamp pace being slower, and even if we have more time to teach stuff, I think we shouldn't add to the bootcamp's curriculum.

**Comment by dominguezmaria** [comment]
@PedroGFonseca [suggestion](https://github.com/LDSSA/wiki/issues/47): add [DummyClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.dummy.DummyClassifier.html) to the Workflow SLU.

**Comment by dominguezmaria** [comment]
@rjna [suggestion](https://github.com/LDSSA/wiki/issues/46): add Undersampling and oversampling to the curriculum

**Comment by CatarinaSilva** [comment]
hey @dominguezmaria I let's sett the deadline to the 20th of April for the bootcamp curriculum to be closed so we can start. Can I assume the 3 first BLUs did not change so I already get people to work on them. I'm still putting together the final proposal regarding deadlines for teaching but for these three I would like instructors on it already in April, if possible.

Regarding the suggestions:

> SVMs would go to Text Classification 

I think it might feel disconnected from the rest but let's maybe open an issue to discuss spec 4's curriculum? 

> kNN would go to RecSys

Not sure about this one, @ivopbernardo does it make sense to you?

> Tree Based models would still be in the Bootcamp (but as optional content).
> (and regarding optional content)

I would say you could still have the presentations but just not have them as hard requirements to move forward. In this sense I would reorganize anything optional to be presented last on either of the bootcamp days. I do agree with Pedro that tree based methods seem important. Any other more advanced models I agree with making optional.

> Undersampling and oversampling

Could we add this to feature engineering or is it too big? or to Data Sufficiency & Selection?

Finally, just to confirm, this is what we have so far on the table:

### Admissions

U | Name | Core Competency | Status 
-- | -- | -- | -- 
SLU01 | Pandas 101 | Data Skills & Analysis | Keep 
SLU02 | Subsetting Data in Pandas | Data Skills & Analysis | Keep 
SLU03 | Visualization with Pandas & Matplotlib | Data Skills & Analysis | Keep 

### Bootcamp

U | Name | Core Competency | Status 
-- | -- | -- | -- 
SLU04 | Basic Stats with Pandas | Data Skills & Analysis | Keep
SLU05 | Covariance & Correlation | Data Skills & Analysis | Keep
SLU06 | Dealing with Data Problems | Data Skills & Analysis | Keep
SLU07 | Regression with Linear Regression | Learning Algorithms | Keep
SLU08 | Classification with Logistic Regression | Learning Algorithms | Keep
SLU09 | Model Selection & Overfitting | ML Fundamentals | Keep
SLU10 | Metrics for Regression | ML Fundamentals | Keep
SLU11 | Metrics for Classification | ML Fundamentals | Keep
SLU12 | Support Vector Machines (SVM) | Learning Algorithms | Move to Spec 4
SLU13 | Tree-Based Models | Learning Algorithms | Keep
SLU14 | k-Nearest Neighbors (kNN) | Learning Algorithms | Move to Spec 5
SLU15 | Feature Engineering (aka Real Wold Data) | Data Skills & Analysis  | Keep and potentially add Under/Oversampling?
SLU16 | Data Sufficiency & Selection | Data Skills & Analysis  | Keep
SLU17 | Ethics & Fairness | Data Skills & Analysis | Keep
SLU18 | Hyperparameter Tuning | ML Fundamentals  | Keep
SLU19 | Workflow | | Keep but add DummyClassifier 


Is this correct?

If so, with 2 less units and one more week I think it should already be lighter on the students, but we could still decide if we want to make any of the units "optional", as in they don't need to pass it to proceed

**Comment by ivopbernardo** [comment]
hey @CatarinaSilva and @dominguezmaria !

I am not sure how we could fit KNN in RecSys, at least with how we have the materials at the moment. We never approach the concept of binary targets in the context of RecSys due to the new concepts being introduced (sparse matrices, collaborative filtering) that tend to cause some struggle for students.

We would need to find an "angle" on how we can introduce a binary problem in this approach - we could use KNN to solve the Cold Start problem but I can't seem to grasp what would be the "real" labels we would like target. I actually tend to solve those problems with an unsupervised approach like K-Means.

Any ideas?

**Comment by dominguezmaria** [comment]
On the optional learning units, I'm in favour of removing them from the presentations. The amount of information that we present in the bootcamp is still huge and I believe that students will only absorb the largest part of the content when they're going over it at their pace, later on.
I'm up for debate on this one, but I don't think it's that valuable to have a presentation about a topic that we'll not encourage them to revisit later on.

The reason why I suggested to have SVMs and kNN as optional was because we have the materials, so if people are curious and wanting to explore further, they would have them there.

The reason for my suggestion of moving SVMs to Text Classification and kNN to RecSys is that we could present these algorithms together with examples of their usage.
I checked all the LUs and the only place where we're using an SVM based model is in the Text Classification spec. 
And @ivopbernardo, we don't need to find angles to justify why kNN (as an algorithm, not as a classifier) should be presented in the RecSys specialisation: collaborative filtering is a kNN based model!

Thoughts?

**Comment by ivopbernardo** [comment]
The angle would be needed if you wanted to include the classifier somewhere during the RecSys materials - we completely don't advise treating Rec. as a binary problem ahah Otherwise, if that's not the case and the optional part is a completely different notebook separated from the main materials, alrighty then!

**Comment by dominguezmaria** [comment]
> The angle would be needed if you wanted to include the classifier somewhere during the RecSys materials - we completely don't advise treating Rec. as a binary problem ahah Otherwise, if that's not the case and the optional part is a completely different notebook separated from the main materials, alrighty then!

I'm referring to kNN as an algorithm **not as a classifier**.
kNN is an algorithm where, to make a prediction, you select a neighbourhood around a point, pick its k closest neighbours, combine the neighbours' features somehow and output a prediction. You can use this idea to implement a kNN classifier, a kNN regressor or a "kNN recommender" (which is basically what collaborative filtering is!).
In fact, I think that kNN is so relevant in this topic that we should even include in the recsys curriculum (not only as optional materials).

**Comment by ivopbernardo** [comment]
I think I am not being clear and probably being lost in my own semantic, sorry. When I'm referring to KNN I'm always assuming the **KNN Classifier** that was the focus of the application of KNN in the SLU, not the general KNN approach. The NN approach is already included and mentioned in the RecSys materials (and probably can be improved) - what is not included and I don't agree that is, is the KNN Classifier (that was the focus of the application of the SLU) - that can be optional. 

We're probably saying the same thing, but you are way more precise than I am!

**Comment by CatarinaSilva** [comment]
@dominguezmaria we use SVMs but the classifier there is not really the point. That being said, we could, however, create additional readings that are not essential to pass the exercises but that are provided for whoever is interested (as we do with going deeper in the chi-squared math, for example). Would that make sense?

Btw, for the units moved, they don't become optional in the bootcamp, right? They become part of the specialization 

Which other units would you see as optional, if any?

**Comment by dominguezmaria** [comment]
> I think I am not being clear and probably being lost in my own semantic, sorry. When I'm referring to KNN I'm always assuming the **KNN Classifier** that was the focus of the application of KNN in the SLU, not the general KNN approach. The NN approach is already included and mentioned in the RecSys materials (and probably can be improved) - what is not included and I don't agree that is, is the KNN Classifier (that was the focus of the application of the SLU) - that can be optional.
> 
> We're probably saying the same thing, but you are way more precise than I am!

Ahh I see what you're saying! In my mind, the SLU's focus was not that much on the classifier, but more on the algorithm itself, and then the classifier was an example application of the algorithm.
In that case, I'm for removing this SLU from the bootcamp and only talking about nearest neighbours in the RecSys sepcialisation.

**Comment by dominguezmaria** [comment]
> @dominguezmaria we use SVMs but the classifier there is not really the point. That being said, we could, however, create additional readings that are not essential to pass the exercises but that are provided for whoever is interested (as we do with going deeper in the chi-squared math, for example). Would that make sense?
> 
> Btw, for the units moved, they don't become optional in the bootcamp, right? They become part of the specialization
> 
> Which other units would you see as optional, if any?

Yes, I was thinking that the materials that we already have on SVMs would go in like the chi-squared math: for whoever is interested only. I was suggesting in the Text Classification specialisation because we happen to use it there (even if it's not that important that it's an SVM based classifier).
But I'm also happy to just remove the SVM content all together.

The other content I was suggesting having as optional was the content in the kNN SLU. I suggested moving it into the RecSys specialisation because we talk about nearest neighbours there. But as @ivopbernardo pointed out, we're already explaining the kNN algorithm there, so it may feel a bit out of scope to be mentioning the kNN classifier or regressor at that point.

But we could also have these as optional material (without presentation and evaluation not mandatory) in the bootcamp, like extra SLUs about ML algorithms.

**Comment by CatarinaSilva** [comment]
Yap I think keeping it "in the bootcamp" but as optional can be good. No presentation, just the notebooks, and they are not mandatory 😄 

What do you think of adding the undersampling/oversampling to either "Feature Engineering" (SLU15) or "Data Sufficiency & Selection" (SLU16)? Or should they be a new SLU, and if so, an optional one?

The same goes for dummy classifiers, where should they go?

**Comment by CatarinaSilva** [comment]
Hey @dominguezmaria so do we have a final version of this?

A couple of questions left responding:
* Moving SLU12 and 14 to other specs or making them optional?
* what are the requirements for passing ?
* what is the distribution over the 2 days of bootcamp? 
* Where do the new topics go (undersampling/oversampling + dummy classifiers)?

If I may give some suggestions for the curriculum:

* Making SVMs and KNN classifiers optional units instead of moving to a spec (and we could even think if there are other models that can be nice to add as optional units)
* Presenting both metrics SLUs (SLU 10 and 11) before model selection and overfitting (also did this so the days had a nice consistent flow)
* Dummy classifiers could go in the workflow unit
* Under and oversampling could be mentioned on data sufficiency and selection or made another optional unit with more depth (or both 😂 )

And for the requirements, I would suggest lowering the bar to continue the academy but keeping the bar for the "Pass" in the curriculum:
* Requirement to continue with the academy: get 12/20 in all mandatory units
* Requirement to get a Pass on the final curriculum on the bootcamp: get 16/20 in all mandatory units
* We could also have in the CV a slot for optional bootcamp units where they would get a pass if they deliver them with 16/20 also (we would need to check deadline here though)

[For this requirement part let me know if we have a different issue or this one is fine. I would also like to get rules nailed down quickly so we can update the wiki]

So overall what I'm suggesting boils down to:

### Admissions

U | Name | Core Competency | Required pass (above 16/20)
-- | -- | -- | -- 
SLU01 | Pandas 101 | Data Skills & Analysis | ☑️
SLU02 | Subsetting Data in Pandas | Data Skills & Analysis | ☑️
SLU03 | Visualization with Pandas & Matplotlib | Data Skills & Analysis | ☑️

### Bootcamp - day 1 

U | Name | Core Competency | Required pass (above 16/20) 
-- | -- | -- | -- 
SLU04 | Basic Stats with Pandas | Data Skills & Analysis  | ☑️
SLU05 | Covariance & Correlation | Data Skills & Analysis  | ☑️
SLU06 | Dealing with Data Problems | Data Skills & Analysis  | ☑️
SLU07 | Regression with Linear Regression | Learning Algorithms  | ☑️
SLU08 | Classification with Logistic Regression | Learning Algorithms  | ☑️
SLU10 | Metrics for Regression | ML Fundamentals  | ☑️
SLU11 | Metrics for Classification | ML Fundamentals  | ☑️

### Bootcamp - day 2
U | Name | Core Competency | Required pass (above 16/20) 
-- | -- | -- | -- 
SLU09 | Model Selection & Overfitting | ML Fundamentals  | ☑️
SLU13 | Tree-Based Models | Learning Algorithms  | ☑️
SLU15 | Feature Engineering (aka Real Wold Data) | Data Skills & Analysis  | ☑️
SLU16 | Data Sufficiency & Selection | Data Skills & Analysis  | ☑️
SLU17 | Ethics & Fairness | Data Skills & Analysis | ☑️
SLU18 | Hyperparameter Tuning | ML Fundamentals  | ☑️
SLU19 | Workflow | | ☑️


### Optional notebooks
U | Name | Core Competency | Required pass (above 16/20) 
-- | -- | -- | -- 
SLU12 | Support Vector Machines (SVM) | Learning Algorithms | ❌
SLU14 | k-Nearest Neighbors (kNN) | Learning Algorithms | ❌

**Comment by dominguezmaria** [comment]
Sorry for my delay on this @CatarinaSilva !

**On the LUs about models (kNN, SVMs)**, let's make them optional and introduce them at bootcamp time.
Being optional means that:
* There will be no presentation for them, the contents will just be in the repo
* The contents should be the same as for the other SLUs (learning, example and exercises notebook)
* The students are not required to do anything on these SLUs (they don't need to read them, they don't need to do the exercises)

**On the optional units on the CV**, I absolutely agree that they should be there 👍 (don't know how to action this, who's in charge of the CV?)

**On the requirements for passing**, I understand your argument for lowering the bar for passing.
However, I'm conscious about the bootcamp becoming too easy: this year we have one extra week for the bootcamp contents and we're removing content. I don't want students to feel like they'll die right from the start, but I like that the beginning is tough, to set the right motto for the course 😏 Also simpler rules are easier to understand. So my proposal is that we keep with the 16/20 requirement in the mandatory SLUs both for continuing and for passing.

**On DummyClassifiers**, workflow unit looks good to me.

**On the "Under and oversampling"**, this is what's on my mind. SLU09 - Model Selection & Overfitting covers "Generalization Error", "Model Selection" and "Regularized Linear Regression". The first topic is about identifying overfitting; the second topic is about evaluating your model in a principled way (so that your metrics don't seem better than they are); the third topic is about one way of avoiding overfitting. Under and oversampling are also a way of dealing with the bias/variance tradeoff, so it feels natural to me to place these concepts together.
Then SLU16 - Data Sufficiency & Selection, is again about the bias/variance tradeoff, and how we can discard _features_ in order to have simpler models. So, I'm trying to think how to better organise these concepts... This is where I'm currently at (let me know your thoughts):
* ...
* SLU07 - Regression with Linear Regression
* SLU10 | Metrics for Regression
* SLU08 | Classification with Logistic Regression
* SLU11 | Metrics for Classification
* SLU13 | Tree-Based Models: here we explain the algorithm and say that these algorithms can be used both for classification and regression
* SLU15 | Feature Engineering (aka Real Wold Data): ML models handle numbers, not strings, categories, etc, so let's clean up our data!
* SLU09 | Model Selection & Overfitting: bias/variance tradeoff, model evaluation (train-test split, cross validation) using learning curves to detect overfitting
* SLU16 | Data Sufficiency & Selection: how to deal with overfitting (regularisation - on linear models and tree based models, over and under sampling, discarding features)
* ...

@CatarinaSilva let me know your thoughts.
@PedroGFonseca if you could share your thoughts too!!

**Comment by CatarinaSilva** [comment]
@dominguezmaria agreed with all, in particular the reshuffling of regressions/metrics/model selection.

But let's wait for @PedroGFonseca 's feedback also

**Comment by dominguezmaria** [comment]
Ah forgot to mention that I'd like to remove the Tidy Data section from [SLU06](https://github.com/LDSSA/curriculum-development/blob/master/curriculum/01-bootcamp-and-binary-classification.md#slu06), as it feels a bit hard to explain it and not so useful 😅

**Comment by CatarinaSilva** [comment]
Hi @dominguezmaria how should we proceed with this? Are you going to merge the info into this repo?

Once this is closed, I was thinking if using the outcome from this issue to change the bootcamp page in the wiki so that it is clear about the units/evaluation/what's mandatory and what's not/etc

**Comment by dominguezmaria** [comment]
@CatarinaSilva , created [this PR](https://github.com/LDSSA/curriculum-development/pull/34) with all the changes we've discussed.

Could I get a review? Will merge it after if it looks good to you!
